### PR TITLE
Add compute capacity to trtep engine cache file

### PR DESCRIPTION
**Pull request by yf711**
### Description

Add "_smXX" to trtep engine cache file name, which "sm" stands for "Streaming Multiprocessor".

> The GPU compute capability version is prefixed with "SM" because NVIDIA typically improves and updates the SM in each new GPU architecture.

### Motivation and Context

Github issue: https://github.com/microsoft/onnxruntime/issues/15982

Reduce the chance of misusing incompatible engine cache, when user is switching GPU devices with different compute capacity

* The prevention can't be 100%, as model size & GPU memory size could be another factor to make cache incompatible

### CODE REVIEW

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_utils.h [hunk 1]
```diff
@@ -443,14 +443,23 @@ std::string GetCachePath(const std::string& root, const std::string& name) {
...
+  return compute_capability;
+}
+
 /*
  * Get Timing by compute capability
  *
  */
 std::string GetTimingCachePath(const std::string& root, cudaDeviceProp prop) {
   // append compute capability of the GPU as this invalidates the cache and TRT will throw when loading the cache
   const std::string timing_cache_name = "TensorrtExecutionProvider_cache_cc" +
```
Comment (tianleiwu) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_utils.h:461]: change _cc to _sm as well?

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 2]
```diff
@@ -1993,15 +1993,20 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<FusedNodeAnd
     // Otherwise engine will be handled at inference time.
     std::unique_ptr<nvinfer1::ICudaEngine> trt_engine;
     std::unique_ptr<nvinfer1::IExecutionContext> trt_context;
+
+    // Name the engine cache based on GPU compute capacity to prevent user from loading an incompatible cache
```
Comment (tianleiwu) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1997]: "partially" prevent

Note: for example engine cache generated for RTX 3090 (24GB memory) might not run in RTX 3050 (6GB memory?) even though they have same compute capactiy.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 3]
```diff
@@ -1993,15 +1993,20 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<FusedNodeAnd
...
+    cudaDeviceProp prop;
+    std::string compute_capability = "";
+    CUDA_CALL_THROW(cudaGetDeviceProperties(&prop, device_id_));
+    compute_capability = GetComputeCapacity(prop);
```
Comment (tianleiwu) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:2001]: nit: Merge two lines like `std::string compute_capability = GetComputeCapacity(prop);

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_utils.h [hunk 4]
```diff
@@ -443,14 +443,23 @@ std::string GetCachePath(const std::string& root, const std::string& name) {
   }
 }
 
+/*
+ * Get compute capability
+ *
+ */
+std::string GetComputeCapacity(cudaDeviceProp prop) {
```
Comment (tianleiwu) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_utils.h:450]: const cudaDeviceProp &

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc (hunk 2 above)
Comment (yf711) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: Thanks for pointing that out! Comments have been updated

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc (hunk 3 above)
Comment (yf711) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: Fixed

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_utils.h (hunk 4 above)
Comment (yf711) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_utils.h]: Fixed

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_utils.h (hunk 1 above)
Comment (yf711) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_utils.h]: Done

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**